### PR TITLE
IGNITE-20125 Write to writable indices when writing to partition

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-20125

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 1]
```diff
@@ -2528,7 +2534,9 @@ private CompletableFuture<CompletableFuture<?>> applyUpdateCommand(
                             true,
                             null,
                             null,
-                            null);
+                            null,
+                            indexIdsAtRwTxBeginTs(catalogService, txId, tableId())
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:2538]: It looks like all `handleUpdate`/`handleUpdateAll` invocations get the same last argument: `indexIdsAtRwTxBeginTs(catalogService, txId, tableId()`.

I suggest to calculate this inside them and remove it as an argument from these 2 methods.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableUtils.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.List;
+import java.util.UUID;
+import org.apache.ignite.internal.catalog.CatalogService;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogObjectDescriptor;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.tx.TransactionIds;
+
+/** Auxiliary class. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableUtils.java:30]: Open to suggestions.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableUtils.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+public class TableUtils {
+    /**
+     * Returns index IDs for the table of interest from the catalog for the active catalog version at the beginning timestamp of read-write
+     * transaction.
+     *
+     * <p>NOTE: To avoid races and errors, it is important to call the this method after schema sync at beginTs or to be sure that the
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableUtils.java:36]: fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexUpdateHandler.java [hunk 4]
```diff
@@ -143,20 +154,58 @@ public void buildIndex(int indexId, Stream<BinaryRowAndRowId> rowStream, @Nullab
...
+    private Iterable<TableSchemaAwareIndexStorage> indexes(@Nullable List<Integer> indexIds) {
+        Map<Integer, TableSchemaAwareIndexStorage> indexStorageById = indexes.get();
+
+        assert !indexStorageById.isEmpty();
+
+        if (indexIds == null) {
+            return indexStorageById.values();
+        }
+
+        return mapIterable(indexIds, indexId -> {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexUpdateHandler.java:179]: What was available =)

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        verify(droppedIndexStorage).put(any());
+
+        assertDoesNotThrow(rwTx::commit);
+    }
+
+    private static IgniteImpl node() {
+        return CLUSTER.node(0);
+    }
+
+    private static void dropBuildAllIndexMessages() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:116]: Yes, you're right, I'll change my name.

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+        return indexStorage;
+    }
+
+    private static int indexId(String indexName) {
+        IgniteImpl node = node();
+
+        return TableTestUtils.getIndexIdStrict(node.catalogManager(), indexName, node.clock().nowLong());
+    }
+}
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:145]: fix it

File: modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    @AfterEach
+    void tearDown() {
+        sql("DROP TABLE IF EXISTS " + TABLE_NAME);
+        sql("DROP ZONE IF EXISTS " + ZONE_NAME);
+
+        CLUSTER.runningNodes().forEach(IgniteImpl::stopDroppingMessages);
+    }
+
+    @Test
+    void testCreateIndexInsideRwTransaction() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/index/src/integrationTest/java/org/apache/ignite/internal/index/ItRwTransactionAndIndexesTest.java:66]: We discussed it personally and decided to leave it as is.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 1 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:2538]: fix it

**Review by rpuch — APPROVED** [review]

**Pull request merged**